### PR TITLE
Add `createGraphClient`

### PR DESCRIPTION
**Pull request by MicroFish91**
Will need to utilize Microsoft Graph to add support for CRUD operations on Service Principals, see: https://github.com/microsoft/vscode-azurecontainerapps/issues/313#issuecomment-1518418209

Shoutout to @nturinski for helping me get unstuck with proper scopes and setup for the client.

Here's how the login process looks the first time `createGraphClient` is called:
image

image

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -4,7 +4,8 @@
         "target": "es6",
         "outDir": "out",
         "lib": [
-            "es6"
+            "es6",
+            "dom"
```
Comment (MicroFish91) [inline tsconfig.json:8]: Graph client uses some DOM type references underneath the hood... this solved the related compile errors I was getting

Comment (alexweininger) [reply tsconfig.json:8]: Oh that's unfortunate. I wish there was a way to only include the exact part of the DOM lib we actually need.

**Review by alexweininger — COMMENTED** [review]
Just for clarity, users have to login to this separately although they might already be logged into Azure since this needs different scopes.

Is it possible to request _all_ the scopes we need including these ones at the beginning or do we want to avoid that?

**Comment by MicroFish91** [comment]
> Just for clarity, users have to login to this separately although they might already be logged into Azure since this needs different scopes.
> 
> Is it possible to request _all_ the scopes we need including these ones at the beginning or do we want to avoid that?

Hmm, yeah I'm curious what the consensus is for this.  I'd have to look into it, but I was originally planning to just go with the defaults for now and add onto the scope as the need arose.  I've only tested so far with the `/me` endpoint, so I don't know which all scopes I will be needing that the defaults don't already cover.

File: tsconfig.json (hunk 1 above)
Comment (bwateratmsft) [reply tsconfig.json:8]: I'd at least leave a comment in the tsconfig about it, but this feels bad.

File: src/utils/graphClient.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { AuthenticationProvider, Client, ClientOptions } from "@microsoft/microsoft-graph-client";
```
Comment (bwateratmsft) [inline src/utils/graphClient.ts:6]:
```suggestion
import { AuthenticationProvider, Client as GraphClient, ClientOptions as GraphClientOptions } from "@microsoft/microsoft-graph-client";
```

And corresponding changes below

**Comment by MicroFish91** [comment]
> Is it plausible to pull it off with direct HTTP requests rather than adding a new dependency (especially with the DOM thing), or is it complicated enough to justify using it?

Sure, let me investigate making the request without the dependency - I'll get back to you on this one

**Comment by alexweininger** [comment]
> > Just for clarity, users have to login to this separately although they might already be logged into Azure since this needs different scopes.
> > Is it possible to request _all_ the scopes we need including these ones at the beginning or do we want to avoid that?
> 
> When you say beginning, are you referring to initializing this client earlier on during like the initial extension activation? Or is the question more focused on maximizing our scopes and we just create the first instance of the client whenever we need to actually use it?

My assumptions are: a new user who wants to use this feature will have to sign in twice. Once to sign into Azure, and again when we need to create this graph client.

My question is: Is it feasible to somehow still only require a single login?

How I _think_ it might be possible: Request the scopes needed to create the graph client on the first sign in, that way we can create the graph client later on without an additional sign in.

With that said, even if we can do that, I'm not sure it's a great idea to request a lot of additional scopes if we don't necessarily plan to always use those scopes. But I think it's worth investigating.

**Comment by MicroFish91** [comment]
Everything works now with the direct HTTP request and with removal of the Microsoft Graph SDK... however, I'm still investigating how to get expand the scopes to operate with Service Principals.

**Comment by MicroFish91** [comment]
Update: Will probably end up closing this out for a different solution... waiting for some final confirmations before doing so

**Pull request closed without merging**